[REEF-184] Allow NetworkService to add more than one message handlers

taegeonum opened the pull request:
This pull request addressed the issue by

  * Introduce an interface of `NetworkServiceClient`
  * Implement `DefaultNetworkServiceImpl` which is an implementation of `NetworkServiceClient`
  * NetworkServiceClient is a singleton.
  * Users register their codec, eventHandler, and linkListener. NetworkServiceClient registers them and creates a ConnectionFactory. `networkService.registerConnectionFactory(clientServiceId, codec,
eventHandler, linkListener)`
  * After the registeration, users can get a ConnectionFactory `<T> ConnectionFactory<T>
networkService.getConnectionFactory(clientServiceId)`
  * Users can call `Connection<T>
connectionFactory.newConnection(Identifier destId)` in order to create a connection
between destId and local. 

Acknowledgement: Collaborates with @gwsshs22 (GeonWoo Kim)

JIRA: [REEF-184](http://issues.apache.org/jira/browse/REEF-184)

gwsshs22 commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSConnection.java:
```diff
@@ -0,0 +1,88 @@
+    link = connFactory.openLink(remoteId);
+  }
+
+  public void write(final List<T> messageList) throws NetworkException {
```
The "throws NetworkException" clause in Link interface was removed by https://issues.apache.org/jira/browse/REEF-182, how about changing Connection interface to remove the throws clause? (It seems to be the write method is called synchronously and throw some exception when an error occurs while writing but actually it works asynchronously.)

bgchun replied:
Can you address @gwsshs22's comment?

taegeonum replied:
Done by 51b3923235cc020e33a0e46d749a6275611a1825

markusweimer commented:
@bgchun @shravanmn @dkm2110 @beysims  either of you available to review this? I can do style checks and such, but this is too important not to be checked by a person with experience in networking.

bgchun commented:
@markusweimer I can review it early next week.

markusweimer commented:
@taegeonum you might want to merge this with current `master` before @bgchun reviews it. We've had quite a few changes since this was posted.

bgchun commented:
@taegeonum Any changes we need to consider before I review this request?

taegeonum commented:
@bgchun No changes. Please review this request. Thanks!

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+
+@DefaultImplementation(DefaultNetworkServiceImpl.class)
+/**
+ * NetworkServiceClient.
```
The description is not self-contained. Can you elaborate?

taegeonum replied:
Done by aefaf22c013558fb7dd7ab57bbb05a0d8efe0226

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+
+import java.net.SocketAddress;
+
+@DefaultImplementation(DefaultNetworkServiceImpl.class)
```
Why not DefaultNetworkServiceClientImpl.class?

taegeonum replied:
Done by 7f56ab5da1ff1405c8e916b7054e2f0e865e3679

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+   * Binds Codec, EventHandler, and LinkListener to the ConnectionFactory.
+   * @throws NetworkException throws a NetworkException when duplicated connectionFactoryId exists.
+   */
+  <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
```
Why is the connection factory ID type Class<? extends Name<String>>?

bgchun replied:
Explain each parameter individually in the comment.

taegeonum replied:
We can register and get a ConnectionFactory instance by using the same instance of `Class<? extends Name<String>>`. I think it can prevent typos compared to just using `String` or `Identifier`. This approach is also used in `CommunicationGroupDriver.addBroadcast`.

bgchun replied:
Why Class<? extends Name<String>>? Why not Class<Name<String>>?
Can you show me examples of using Identifier vs Class<? extends
Name<String>>?


On Sun, Jul 5, 2015 at 10:08 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33890969>:
>
> > +
> > +import java.net.SocketAddress;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + * It can bind multiple ConnectionFactory.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
> > +
> > +  /**
> > +   * Creates an instance of ConnectionFactory corresponding to the connectionFactoryId.
> > +   * Binds Codec, EventHandler, and LinkListener to the ConnectionFactory.
> > +   * @throws NetworkException throws a NetworkException when duplicated connectionFactoryId exists.
> > +   */
> > +  <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
>
> We can register and get a ConnectionFactory instance by using the same
> instance of Class<? extends Name<String>>. I think it can prevent typos
> compared to just using String or Identifier. This approach is also used
> in CommunicationGroupDriver.addBroadcast.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33890969>.
>



-- 
Byung-Gon Chun

taegeonum replied:
I followed `CommunicationGroupDriverImpl.addBroadcast` method. This method gets a `Class<? extends Name<String>> operatorName` parameter to register broadcast. 
Users can get this registered broadcast from `communicationGroupClient.getBroadcastSender(Class<? extends Name<String>> operatorName)` method. 

This could be `Identifier`. Is it better to use `Identifier`?

bgchun replied:
Did we talk about a case where there can be multiple connection factories
for a type? We probably did. Can you double check?


On Sun, Jul 5, 2015 at 11:11 PM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33895118>:
>
> > +
> > +import java.net.SocketAddress;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + * It can bind multiple ConnectionFactory.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
> > +
> > +  /**
> > +   * Creates an instance of ConnectionFactory corresponding to the connectionFactoryId.
> > +   * Binds Codec, EventHandler, and LinkListener to the ConnectionFactory.
> > +   * @throws NetworkException throws a NetworkException when duplicated connectionFactoryId exists.
> > +   */
> > +  <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
>
> I followed CommunicationGroupDriverImpl.addBroadcast method. This method
> gets a Class<? extends Name<String>> operatorName parameter to register
> broadcast.
> Users can get this registered broadcast from communicationGroupClient.getBroadcastSender(Class<?
> extends Name<String>> operatorName) method.
>
> This could be Identifier. Is it better to use Identifier?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33895118>.
>



-- 
Byung-Gon Chun

bgchun replied:
It doesn't sound like a good argument to say you followed the addBroadcast
method.
Can you compare the approach that uses Class<? extends Name<String>> with
one that uses Identifier? What're pros and cons?

On Sun, Jul 5, 2015 at 11:11 PM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33895118>:
>
> > +
> > +import java.net.SocketAddress;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + * It can bind multiple ConnectionFactory.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
> > +
> > +  /**
> > +   * Creates an instance of ConnectionFactory corresponding to the connectionFactoryId.
> > +   * Binds Codec, EventHandler, and LinkListener to the ConnectionFactory.
> > +   * @throws NetworkException throws a NetworkException when duplicated connectionFactoryId exists.
> > +   */
> > +  <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
>
> I followed CommunicationGroupDriverImpl.addBroadcast method. This method
> gets a Class<? extends Name<String>> operatorName parameter to register
> broadcast.
> Users can get this registered broadcast from communicationGroupClient.getBroadcastSender(Class<?
> extends Name<String>> operatorName) method.
>
> This could be Identifier. Is it better to use Identifier?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33895118>.
>



-- 
Byung-Gon Chun

taegeonum replied:
1. Yes, we can bind multiple connection factories for a type, by using `registerConnectionFactory`. 

2. `.registerConnectionFactory` and `.getConnectionFactory` can be called in different classes.
If we use Identifier , users have to create the same identifier whenever they want to call `.registerConnectionFactory` and `.getConnectionFactory`.
However, if we use `Class<? extends Name<String>>`, users only have to create the class and use it. I think it is convenient for users to use `Class<? extends Name<String>>` compared to `Identifier`.
Is it reasonable?

taegeonum replied:
In other words, `Class<? extends Name<String>>` approach is good when a user registers  static id of `ConnectionFactory`, and `Identifier` is good when a user registers dynamic id of `ConnectionFactory`. I assume that a user want to use static id of `ConnectionFactory`.

bgchun replied:
@taegeonum I don't think it's convenient for users to use `Class<? extends Name<String>>`. This type does not convey anything about the variable is an identifier. 
@dafrista and I also discussed offline and we think that it's better to use `Identifier`. Can you change it to `Identifier`?

taegeonum replied:
Sure. I will change it.

taegeonum replied:
Done by b9a3d24

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+  <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
+                                            final Codec<T> codec,
+                                            final EventHandler<Message<T>> eventHandler,
+                                            final LinkListener<Message<T>> linkListener) throws NetworkException;
```
When does it raise NetworkExceptions?

taegeonum replied:
It raises if the connectionFactoryId was already registered.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+
+  /**
+   * Unregisters connectionFactory
+   * @param connectionFactoryId
```
Description missing. Fix all the comments in this interface.

taegeonum replied:
Done.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+   * Register task identifier.
+   * @param nsId
+   */
+  void registerId(final Identifier nsId);
```
It'd be good to add a description on why we need this interface.

taegeonum replied:
Done.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,82 @@
+  /**
+   * Get local address
+   */
+  SocketAddress getLocalAddress();
```
How is this method used by the client of this interface?

taegeonum replied:
I think it doesn't have to be in this interface. I will remove it. Thanks!

taegeonum replied:
Done by aefaf22c013558fb7dd7ab57bbb05a0d8efe0226.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/BindNSClientToTask.java:
```diff
@@ -0,0 +1,46 @@
+
+import javax.inject.Inject;
+
+public final class BindNSClientToTask implements EventHandler<TaskStart> {
```
Class description missing.

bgchun replied:
The class name is misleading. 
One would expect all the initialization to be done in TaskStart since it implements EventHandler<TaskStart>. 
What if we have other initialization to do in the future? We need to change the name.

taegeonum replied:
I've added the description via 2d06074. This is a class for registering NetworkServiceClient id, which is similar to `lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/BindNSToTask.java`

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSConnection.java:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  @Override
+  public void write(final T message) throws NetworkException {
```
Later we may want to introduce library-level batching. We don't want to send each message separately.

taegeonum replied:
I've added `write(List<T> messages)` into the `Connection` interface.

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSConnection.java:
```diff
@@ -0,0 +1,88 @@
+  public void close() {
+    if (closed.compareAndSet(false, true)) {
+      connFactory.removeConnection(this.remoteId);
+      link = null;   // TODO: Link should be closed. Current link interface does not have .close() method.
```
No. Link should not be closed by the user. Remove this TODO.

taegeonum replied:
Done.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSConnection.java:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  @Override
+  public String toString() {
```
Use StringBuilder!

taegeonum replied:
Done.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSConnection.java:
```diff
@@ -0,0 +1,88 @@
+        + connFactory.getConnectionFactoryId() + " to "
+        + remoteId + ":" + connFactory.getConnectionFactoryId();
+  }
+}
```
What's this character at the end of the file?

taegeonum replied:
I'm not sure. I didn't do anything.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessage.java:
```diff
@@ -0,0 +1,141 @@
+import java.util.List;
+
+
+final class DefaultNSMessage<T> implements Message<T> {
```
Comment missing.

taegeonum replied:
Done.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessage.java:
```diff
@@ -0,0 +1,141 @@
+   * @param remoteId   the remote identifier
+   * @param messages  the list of messages
+   */
+  public DefaultNSMessage(
```
We need to support StreamCodec similar to .NET.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessage.java:
```diff
@@ -0,0 +1,141 @@
+   * @param messages  the list of messages
+   */
+  public DefaultNSMessage(
+      final String connectionFactoryId,
```
Inconsistent naming.
connFactory and connectionFactory are mixed up.

taegeonum replied:
I've changed all of the connectionFactory to connFactory

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessage.java:
```diff
@@ -0,0 +1,141 @@
+   */
+  public DefaultNSMessage(
+      final String connectionFactoryId,
+      final Identifier srcId,
```
srcId or sourceId ?

taegeonum replied:
I decided to keep srcId because in `Message` interface, it also uses `srcId`

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessage.java:
```diff
@@ -0,0 +1,141 @@
+   */
+  public T getDataAt(final int index) {
+    if (index >= messages.size()) {
+      return null;
```
Can we use Optional?

taegeonum replied:
I've removed this method, because it is unnecessary. Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessage.java:
```diff
@@ -0,0 +1,141 @@
+   *
+   * @return a size of messages
+   */
+  public int getDataSize() {
```
Bad naming.

taegeonum replied:
I've removed this method, because it is unnecessary. Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,133 @@
+
+/**
+ * NSMessage codec.
+ *
```
Fix the comment.

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,133 @@
+final class DefaultNSMessageCodec implements Codec<DefaultNSMessage> {
+
+  private final IdentifierFactory factory;
+  private final Map<String, NSConnectionFactory> connectionFactoryMap;
```
Add a comment on what this map maintains.

taegeonum replied:
Done by 9f48a5be9b5d4ef8385cd70b2d09aed354ae14ea

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,133 @@
+
+  private final IdentifierFactory factory;
+  private final Map<String, NSConnectionFactory> connectionFactoryMap;
+  private final Map<String, Boolean> isStreamingCodecMap;
```
Add a comment on what this map maintains.

taegeonum replied:
Done by 9f48a5be9b5d4ef8385cd70b2d09aed354ae14ea

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,133 @@
+          }
+        } else {
+          final Iterable dataList = obj.getData();
+          for (final Object message : dataList) {
```
Any reason not to use StreamingCodec?

taegeonum replied:
Because users can use default `Codec` interface.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,133 @@
+  DefaultNSMessageCodec(
+      final IdentifierFactory factory,
+      final Map<String, NSConnectionFactory> connectionFactoryMap,
+      final Map<String, Boolean> isStreamingCodecMap) {
```
Don't like to maintain a separate map for each data member. 
Why don't we maintain a map that has <String, DataX> where DataX has all the relevant members?

taegeonum replied:
I've create the streamingCodecMap within DefaultNSMessageCodec. Please look at 9f48a5be9b5d4ef8385cd70b2d09aed354ae14ea

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,133 @@
+        );
+      }
+    } catch (final IOException e) {
+      throw new RuntimeException("IOException", e);
```
This makes me a bit uncomfortable.

taegeonum replied:
Can you please explain why it is uncomfortable?

bgchun replied:
RuntimeException is not a checked exception...

On Sun, Jul 5, 2015 at 11:10 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33891386>:
>
> > +            final int byteSize = dais.readInt();
> > +            final byte[] bytes = new byte[byteSize];
> > +            dais.read(bytes);
> > +            list.add(codec.decode(bytes));
> > +          }
> > +        }
> > +
> > +        return new DefaultNSMessage(
> > +            connectionFactoryId,
> > +            srcId,
> > +            destId,
> > +            list
> > +        );
> > +      }
> > +    } catch (final IOException e) {
> > +      throw new RuntimeException("IOException", e);
>
> Can you please explain why it is uncomfortable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33891386>.
>



-- 
Byung-Gon Chun

bgchun replied:
There is nothing we can do about it at the moment.

On Sun, Jul 5, 2015 at 11:10 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33891386>:
>
> > +            final int byteSize = dais.readInt();
> > +            final byte[] bytes = new byte[byteSize];
> > +            dais.read(bytes);
> > +            list.add(codec.decode(bytes));
> > +          }
> > +        }
> > +
> > +        return new DefaultNSMessage(
> > +            connectionFactoryId,
> > +            srcId,
> > +            destId,
> > +            list
> > +        );
> > +      }
> > +    } catch (final IOException e) {
> > +      throw new RuntimeException("IOException", e);
>
> Can you please explain why it is uncomfortable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33891386>.
>



-- 
Byung-Gon Chun

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+/**
+ * Default NetworkServiceClient implementation
+ */
+public final class DefaultNetworkServiceImpl implements NetworkServiceClient {
```
DefaultNetworkServiceImpl or DefaultNetworkServiceClientImpl ?

taegeonum replied:
I've changed to DefaultNetworkServiceClientImpl

taegeonum replied:
Done by 7f56ab5

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+
+  private static final Logger LOG = Logger.getLogger(DefaultNetworkServiceImpl.class.getName());
+
+  private final IdentifierFactory idFactory;
```
Can you add comments to the private members? This would be nice.

taegeonum replied:
Done by baf310c

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  @Override
+  public synchronized <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
```
Only registerConnectionFactory and unregisterConnectionFactory are synchronized.
We don't care about the behavior of other connection factory methods (e.g., getConnctionFactory)?

taegeonum replied:
`getConnectionFactory` is just to read map. I don't think we have to `synchronize` this method.

bgchun replied:
Well, it can read inconsistent state from the map while it's being
modified.

On Sun, Jul 5, 2015 at 11:03 PM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33895084>:
>
> > +      @Override
> > +      public void onNext(final Identifier id) {
> > +        try {
> > +          nameResolver.unregister(id);
> > +          LOG.log(Level.FINEST, "Unregistered {0} with nameservice", id);
> > +        } catch (final Exception ex) {
> > +          final String msg = "Unable to unregister " + id + " with name service";
> > +          LOG.log(Level.WARNING, msg, ex);
> > +          throw new RuntimeException(msg, ex);
> > +        }
> > +      }
> > +    }, 5);
> > +  }
> > +
> > +  @Override
> > +  public synchronized <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
>
> getConnectionFactory is just to read map. I don't think we have to
> synchronize this method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33895084>.
>



-- 
Byung-Gon Chun

bgchun replied:
If performance matters, use ConcurrentMap.

On Sun, Jul 5, 2015 at 11:03 PM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33895084>:
>
> > +      @Override
> > +      public void onNext(final Identifier id) {
> > +        try {
> > +          nameResolver.unregister(id);
> > +          LOG.log(Level.FINEST, "Unregistered {0} with nameservice", id);
> > +        } catch (final Exception ex) {
> > +          final String msg = "Unable to unregister " + id + " with name service";
> > +          LOG.log(Level.WARNING, msg, ex);
> > +          throw new RuntimeException(msg, ex);
> > +        }
> > +      }
> > +    }, 5);
> > +  }
> > +
> > +  @Override
> > +  public synchronized <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
>
> getConnectionFactory is just to read map. I don't think we have to
> synchronize this method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33895084>.
>



-- 
Byung-Gon Chun

taegeonum replied:
Thanks. Actually, the map is ConcurrentMap. I think it would be better to remove synchronized and just use ConcurrentMap. Done by b1d6ad3

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    connectionFactoryMap.put(connFactoryId, new NSConnectionFactory<>(this, connFactoryId, codec, eventHandler, linkListener));
+    isStreamingCodecMap.put(connFactoryId, codec instanceof StreamingCodec);
```
It looks like we don't need to maintain this map separately. We can get the info from connectionFactoryMap.

taegeonum replied:
get rid of it via 8f0a414c4ad5bb70507ae73bdcde5703560d329e, and the isStreamingCodecMap is created within NSMessageCodec via 9f48a5be9b5d4ef8385cd70b2d09aed354ae14ea

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+    if (connFactory != null) {
+      connectionFactoryMap.remove(connFactoryId);
+      isStreamingCodecMap.remove(connFactoryId);
+    }
```
Do logging when connFactory is null.

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+      @Parameter(NetworkServiceParameters.NetworkServicePort.class) final int nsPort,
+      final TransportFactory transportFactory,
+      final NameResolver nameResolver) {
+
```
remove the empty line

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+
+  @Override
+  public <T> ConnectionFactory<T> getConnectionFactory(final Class<? extends Name<String>> connectionFactoryId) {
+
```
remove the empty line

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+  @Override
+  public void unregisterId(final Identifier nsId) {
+    this.myId = null;
+    LOG.log(Level.FINEST, "Unbinding {0} to NetworkService@({1})",
```
Log in the beginning.

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java:
```diff
@@ -0,0 +1,217 @@
+    LOG.log(Level.FINE, "Shutting down");
+    this.nameServiceRegisteringStage.close();
+    this.nameServiceUnregisteringStage.close();
+    this.transport.close();
```
What if this transport is shared? 
The transport should do a reference counting and perform actual close when no one is using it.

taegeonum replied:
`DefaultNetworkServiceImpl` uses `TransportFactory` to create a new `transport`, so it cannot be shared.

bgchun replied:
Right now, it's so. But when we have a version of TransportFactory that
returns the same transport object, it's going to be shared.

On Sun, Jul 5, 2015 at 10:40 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33891141>:
>
> > +  @Override
> > +  public Identifier getNetworkServiceId() {
> > +    return this.myId;
> > +  }
> > +
> > +  @Override
> > +  public SocketAddress getLocalAddress() {
> > +    return transport.getLocalAddress();
> > +  }
> > +
> > +  @Override
> > +  public void close() throws Exception {
> > +    LOG.log(Level.FINE, "Shutting down");
> > +    this.nameServiceRegisteringStage.close();
> > +    this.nameServiceUnregisteringStage.close();
> > +    this.transport.close();
>
> DefaultNetworkServiceImpl uses TransportFactory to create a new transport,
> so it cannot be shared.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33891141>.
>



-- 
Byung-Gon Chun

taegeonum replied:
I agreed. Do you mean that the `transport` should do reference counting in the future? or does this should be implemented in `DefaultNetworkServiceImpl`?

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NSConnectionFactory.java:
```diff
@@ -0,0 +1,101 @@
+/**
+ * A connection factory which is created by NetworkService.
+ */
+
```
Remove the empty line.

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkServiceLinkListener.java:
```diff
@@ -0,0 +1,52 @@
+import java.net.SocketAddress;
+import java.util.Map;
+
+
```
Remove the empty line.

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkServiceReceiveHandler.java:
```diff
@@ -0,0 +1,51 @@
+final class NetworkServiceReceiveHandler implements EventHandler<TransportEvent> {
+
+  private final Map<String, NSConnectionFactory> connectionFactoryMap;
+  private final Codec<DefaultNSMessage> nsEventCodec;
```
A better name?

taegeonum replied:
Changed to codec

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkServiceReceiveHandler.java:
```diff
@@ -0,0 +1,51 @@
+
+  @Override
+  public void onNext(final TransportEvent transportEvent) {
+    final DefaultNSMessage decodedEvent = nsEventCodec.decode(transportEvent.getData());
```
A better name than decodedEvent?

taegeonum replied:
Changed to nsMessage

taegeonum replied:
Done by 8f0a414c4ad5bb70507ae73bdcde5703560d329e

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/UnbindNSClientFromTask.java:
```diff
@@ -0,0 +1,46 @@
+
+import javax.inject.Inject;
+
+public final class UnbindNSClientFromTask implements EventHandler<TaskStop> {
```
Comment missing.
UnbindNSClientFromTask is misleading. It can be part of clean up in the future.

taegeonum replied:
Done by 2d06074

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+
+    LOG.log(Level.FINEST, "=== Test network service receiver start");
+    // network service for task2
+    final Injector ij2 = injector.forkInjector(netConf);
```
ij2 -> injectorTask2

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+
+    // network service for task1
+    LOG.log(Level.FINEST, "=== Test network service sender start");
+    final Injector ij3 = injector.forkInjector(netConf);
```
ij3 -> injectorTask1

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+    final Injector ij2 = injector.forkInjector(netConf);
+    final NetworkServiceClient ns2 = ij2.getInstance(NetworkServiceClient.class);
+    final IdentifierFactory factory = ij2.getNamedInstance(NetworkServiceParameters.NetworkServiceIdentifierFactory.class);
+    ns2.registerId(factory.getNewInstance("task2"));
```
Introduce a variable for "task2"

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+    LOG.log(Level.FINEST, "=== Test network service sender start");
+    final Injector ij3 = injector.forkInjector(netConf);
+    final NetworkServiceClient ns1 = ij3.getInstance(NetworkServiceClient.class);
+    ns1.registerId(factory.getNewInstance("task1"));
```
Similarly introduce a variable for "task1"

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+  }
+
+  /**
+   * Multiple connection factories test
```
Change the comment to reflect actual tests.

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+
+    LOG.log(Level.FINEST, "=== Test network service receiver start");
+    // network service for task2
+    final Injector ij2 = injector.forkInjector(netConf);
```
injector naming change. Please see the comments above.

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,609 @@
+    final ConnectionFactory<Integer> shuffleSender = ns1.getConnectionFactory(ShuffleClientId.class);
+
+    final Identifier destId = factory.getNewInstance("task2");
+    final Connection<String> conn = gcSender.newConnection(destId);
```
conn, conn2 -> use better names

taegeonum replied:
Done by f346171eeedb69b8537aa8c1527bb4ac1c129e33

bgchun commented:
@taegeonum I made a pass. There is no test that uses StreamingCodec. Could you add one?

bgchun commented:
@taegeonum Also, can you elaborate the summary of the pull request? It does not convey well how the issue is addressed.

bgchun commented:
@taegeonum Can you also add a comment about the migration plan (how you will migrate from the old version of NetworkService to the new version)?

bgchun commented:
The server is rebooted, but the windows build still fails.
https://builds.apache.org/job/Reef-pull-request-windows/513/console

On Sun, Jul 5, 2015 at 11:25 AM, asfbot <notifications@github.com> wrote:

> Reef-pull-request-windows #513
> <https://builds.apache.org/job/Reef-pull-request-windows/513/> FAILURE
> Looks like there's a problem with this pull request
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255#issuecomment-118571333>
> .
>



-- 
Byung-Gon Chun

bgchun commented:
The build shows the following error message. Did you run apache rat?

Waiting for Jenkins to finish collecting data[ERROR] Failed to execute
goal org.apache.rat:apache-rat-plugin:0.11:check (validate) on project
reef-project: Too many files with unapproved license: 265 See RAT
report in: /home/jenkins/jenkins-slave/workspace/Reef-pull-request-ubuntu/target/rat.txt
-> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors,
re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X
switch to enable full debug logging.[ERROR] [ERROR] For more
information about the errors and possible solutions, please read the
following articles:[ERROR] [Help 1]
http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException



On Sun, Jul 5, 2015 at 1:48 PM, asfbot <notifications@github.com> wrote:

> Reef-pull-request-ubuntu #676
> <https://builds.apache.org/job/Reef-pull-request-ubuntu/676/> FAILURE
> Looks like there's a problem with this pull request
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255#issuecomment-118579524>
> .
>



-- 
Byung-Gon Chun

taegeonum commented:
@bgchun I did. In my local machine, there is no error message when I run `mvn clean install`. I'm not sure why the rat check shows the error in Jenkins ubuntu. I will send an email to dev@reef.

markusweimer commented:
@taegeonum The build servers merge current `master` into the code before building. And we recently made rat part of the build. If you merge (or rebase onto) master on your side, you should see the same behavior / error.

taegeonum commented:
@markusweimer I've already merged current master into my branch via e9d3e56e00bcdc668e15740bd82e627e4a14ff92.  It looks like there was some problems when I merge the master. 
@bgchun I will rebase current commits and let you know if all of the comments are addressed.

taegeonum commented:
@bgchun I will refactor `GroupComm` to use current `NetworkSeriveClient` and deprecate old `NetworkService` in 0.12, after this pull request is merged. We can remove the old `NetworkService` in 0.13. 

I've addressed all of your comments. Please take a look. Thanks!

bgchun commented:
@taegeonum Could you add the migration plan to JIRA?

taegeonum commented:
@bgchun Sure. I've added a comment to the JIRA.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/Connection.java:
```diff
@@ -35,12 +37,18 @@
 
   /**
-   * Writes an object to the connection.
+   * Writes an message to the connection.
```
an message -> a message

taegeonum replied:
Done by 7d287c9

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,84 @@
+import org.apache.reef.wake.remote.Codec;
+import org.apache.reef.wake.remote.transport.LinkListener;
+
+@DefaultImplementation(DefaultNetworkServiceClientImpl.class)
```
Move this line after the comment.

taegeonum replied:
Done by 7d287c9

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,84 @@
+ * NetworkServiceClient is a service which is designed for communicating messages with each other.
+ * Users can send messages with NetworkServiceClient, by registering their Codec, EventHandler, and LinkListener.
+ */
+public interface NetworkServiceClient extends AutoCloseable {
```
I still think this interface name is confusing. NetworkService acts as both client and server.
We need a better name. Maybe NetworkConnectionProvider or NetworkServiceProvider?

taegeonum replied:
I agreed. Actually, I wanted to take this name as `NetworkService`. But, current `NetworkService` is an implementation class, not an interface. So I couldn't change this name to `NetworkService` because of compatibility. 

Well, it provides `ConnectionFactory`, so how about `NetworkConnectionFactoryProvider`?

taegeonum replied:
In addition to providing `ConnectionFactory`, it behaves like a gateway of transferring messages, and manages `networking`. How about `NetworkServiceManager` or `NetworkManager`?

cmssnubuilder replied:
Current `NetworkService` (implementation class) is in `network.impl` package, and `NetworkServiceClient` is in `network` package. Another option is to change `NetworkServiceClient` interface name to `NetworkService`. 
It could be confusing in 0.12, because `NetworkService` resides in both `network.impl` package and `network` package. But, if the `NetworkService` is the best option for this name, and if we assume that the `network.impl.NetworkService` is removed in 0.13, then it is worth?

taegeonum replied:
Current `NetworkService` (implementation class) is in `network.impl` package, and `NetworkServiceClient` is in `network` package. Another option is to change `NetworkServiceClient` interface name to `NetworkService`. 
It could be confusing in 0.12, because `NetworkService` resides in both `network.impl` package and `network` package. But, if the `NetworkService` is the best option for this name, and if we assume that the `network.impl.NetworkService` is removed in 0.13, then it is worth?

bgchun replied:
Since it does not manage NetworkService's, either NetworkServiceManager or
NetworkManager is not a good one.


On Fri, Jul 10, 2015 at 1:10 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r34272117>:
>
> > +import org.apache.reef.exception.evaluator.NetworkException;
> > +import org.apache.reef.io.network.impl.DefaultNetworkServiceClientImpl;
> > +import org.apache.reef.tang.annotations.DefaultImplementation;
> > +import org.apache.reef.wake.EventHandler;
> > +import org.apache.reef.wake.Identifier;
> > +import org.apache.reef.wake.remote.Codec;
> > +import org.apache.reef.wake.remote.transport.LinkListener;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceClientImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + *
> > + * NetworkServiceClient is a service which is designed for communicating messages with each other.
> > + * Users can send messages with NetworkServiceClient, by registering their Codec, EventHandler, and LinkListener.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
>
> In addition to providing ConnectionFactory, it behaves like a gateway of
> transferring messages, and manages networking. How about
> NetworkServiceManager or NetworkManager?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r34272117>.
>



-- 
Byung-Gon Chun

bgchun replied:
NetworkConnectionFactoryProvider seems to be better.
If we'd like to indicate this is a REEF service, it have to add "Service"
to the name. NetworkServiceProvider? This name doesn't make sense, though.


On Fri, Jul 10, 2015 at 12:59 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r34271034>:
>
> > +import org.apache.reef.exception.evaluator.NetworkException;
> > +import org.apache.reef.io.network.impl.DefaultNetworkServiceClientImpl;
> > +import org.apache.reef.tang.annotations.DefaultImplementation;
> > +import org.apache.reef.wake.EventHandler;
> > +import org.apache.reef.wake.Identifier;
> > +import org.apache.reef.wake.remote.Codec;
> > +import org.apache.reef.wake.remote.transport.LinkListener;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceClientImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + *
> > + * NetworkServiceClient is a service which is designed for communicating messages with each other.
> > + * Users can send messages with NetworkServiceClient, by registering their Codec, EventHandler, and LinkListener.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
>
> I agreed. Actually, I wanted to take this name as NetworkService. But,
> current NetworkService is an implementation class, not an interface. So I
> couldn't change this name to NetworkService because of compatibility.
>
> Well, it provides ConnectionFactory, so how about
> NetworkConnectionFactoryProvider?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r34271034>.
>



-- 
Byung-Gon Chun

taegeonum replied:
Thanks for your suggestion. However, I'm worried that `NetworkConnectionFactoryProvider` seems to limit the behavior of this class. This class not only provides `ConnectionFactory`, but also  register an identifier, in order to connect with each other. This class is responsible for handling encode/decode messages, routing messages to the corresponding `ConnectionFactory`.

Do you think `NetworkService` is not good? how about `NetworkManagerService`,  `NetworkConnectionService` or `NetworkMessagingService`?

bgchun replied:
I thought you used NetworkServiceClient since we cannot use the name
NetworkService. NetworkService is good.

On Fri, Jul 10, 2015 at 5:05 PM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r34334653>:
>
> > +import org.apache.reef.exception.evaluator.NetworkException;
> > +import org.apache.reef.io.network.impl.DefaultNetworkServiceClientImpl;
> > +import org.apache.reef.tang.annotations.DefaultImplementation;
> > +import org.apache.reef.wake.EventHandler;
> > +import org.apache.reef.wake.Identifier;
> > +import org.apache.reef.wake.remote.Codec;
> > +import org.apache.reef.wake.remote.transport.LinkListener;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceClientImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + *
> > + * NetworkServiceClient is a service which is designed for communicating messages with each other.
> > + * Users can send messages with NetworkServiceClient, by registering their Codec, EventHandler, and LinkListener.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
>
> Thanks for your suggestion. However, I'm worried that
> NetworkConnectionFactoryProvider seems to limit the behavior of this
> class. This class not only provides ConnectionFactory, but also register
> an identifier, in order to connect with each other. This class is
> responsible for handling encode/decode messages, routing messages to the
> corresponding ConnectionFactory.
>
> Do you think NetworkService is not good? how about NetworkManagerService,
> NetworkConnectionService or NetworkMessagingService?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r34334653>.
>



-- 
Byung-Gon Chun

taegeonum replied:
We can use the name of `NetworkService` if we add this to a different package. But, it could be confusing in 0.12.

bgchun replied:
Let's go with NetworkConnectionService.

On Fri, Jul 10, 2015 at 5:22 PM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r34335600>:
>
> > +import org.apache.reef.exception.evaluator.NetworkException;
> > +import org.apache.reef.io.network.impl.DefaultNetworkServiceClientImpl;
> > +import org.apache.reef.tang.annotations.DefaultImplementation;
> > +import org.apache.reef.wake.EventHandler;
> > +import org.apache.reef.wake.Identifier;
> > +import org.apache.reef.wake.remote.Codec;
> > +import org.apache.reef.wake.remote.transport.LinkListener;
> > +
> > +@DefaultImplementation(DefaultNetworkServiceClientImpl.class)
> > +/**
> > + * NetworkServiceClient.
> > + *
> > + * NetworkServiceClient is a service which is designed for communicating messages with each other.
> > + * Users can send messages with NetworkServiceClient, by registering their Codec, EventHandler, and LinkListener.
> > + */
> > +public interface NetworkServiceClient extends AutoCloseable {
>
> We can use the name of NetworkService if we add this to a different
> package. But, it could be confusing in 0.12.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r34335600>.
>



-- 
Byung-Gon Chun

taegeonum replied:
Thanks! Changed to NetworkConnectionService by e76ad0a

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,84 @@
+   * @throws NetworkException throws a NetworkException when duplicated connectionFactoryId exists.
+   */
+  <T> void registerConnectionFactory(final Identifier connectionFactoryId,
+                                            final Codec<T> codec,
```
Align arguments (codec, eventHandler, linkListener).

taegeonum replied:
Done by 7d287c9

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,84 @@
+   * @param codec a codec for type <T>
+   * @param eventHandler an event handler for type <T>
+   * @param linkListener a link listener
+   * @throws NetworkException throws a NetworkException when duplicated connectionFactoryId exists.
```
when duplicate connectionFactoryId's exist

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/NetworkServiceClient.java:
```diff
@@ -0,0 +1,84 @@
+   * This can be used for destination identifier
+   * @param nsId network service client identifier
+   */
+  void registerId(final Identifier nsId);
```
nsId or nscId? 
network service client is confusing

taegeonum replied:
Changed to ncsId (NetworkConnectionServiceId) by e76ad0a

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSConnection.java:
```diff
@@ -0,0 +1,96 @@
+  DefaultNSConnection(
+      final NSConnectionFactory connFactory,
+      final Identifier remoteId) {
+
```
Remove the line.

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java:
```diff
@@ -0,0 +1,147 @@
+   */
+  @Override
+  public DefaultNSMessage decode(final byte[] data) {
+
```
Remove this line.

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceClientImpl.java:
```diff
@@ -0,0 +1,238 @@
+/**
+ * Default NetworkServiceClient implementation.
+ */
+public final class DefaultNetworkServiceClientImpl implements NetworkServiceClient {
```
Change the class name.

taegeonum replied:
Done by e76ad0a

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceClientImpl.java:
```diff
@@ -0,0 +1,238 @@
+    if (connFactoryMap.get(id) != null) {
+      throw new NetworkException("ConnectionFactory " + connFactoryId + " was already registered.");
+    }
+   final ConnectionFactory connFactory = connFactoryMap.putIfAbsent(id,
```
indentation fix

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceClientImpl.java:
```diff
@@ -0,0 +1,238 @@
+
+    if (connFactory != null) {
+      throw new NetworkException("ConnectionFactory " + connFactoryId + " was already registered.");
+
```
Remove this line.

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NSConnection.java:
```diff
@@ -100,10 +101,15 @@ public void open() throws NetworkException {
    */
   @Override
-  public void write(final T obj) throws NetworkException {
+  public void write(final T obj) {
```
obj -> message

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NSConnectionFactory.java:
```diff
@@ -0,0 +1,100 @@
+      final Codec<T> eventCodec,
+      final EventHandler<Message<T>> eventHandler,
+      final LinkListener<Message<T>> eventListener) {
+
```
Remove this line.

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,746 @@
+  }
+
+  public static final class ShuffleClientId implements Name<String> {
+
```
Remove this line.

taegeonum replied:
Done by 5c21682

bgchun commented on lang/java/reef-io/src/test/java/org/apache/reef/services/network/DefaultNetworkServiceTest.java:
```diff
@@ -0,0 +1,746 @@
+   * Test NetworkService registering multiple connection factories which contain streamingcodec.
+   */
+  @Test
+  public void testMultipleConnectionFactoriesStreamingTest() throws Exception {
```
Can you refactor testMultipleConnectionFactoriesStreamingTest?
There's duplicate code in testMultipleConnectionFactoriesTest and testMultipleConnectionFactoriesStreamingTest.

taegeonum replied:
Done by fde624d

taegeonum replied:
Done by fde624d

bgchun commented:
@taegeonum I did my pass. I think we need to come up with a better name. NetworkServiceClient seems to be confusing. Talk to youngbin and geonwoo.

bgchun commented:
@taegeonum Let me know when you are ready for another round of code review.

taegeonum commented:
@bgchun I've addressed all of your comments. Please take a look at them. Thanks!

bgchun commented:
@taegeonum It looks good. Can you address the following checkstyle violation quickly so that I can test and merge this?

[INFO] There are 9 errors reported by Checkstyle 6.6 with lang/java/reef-common/src/main/resources/checkstyle.xml ruleset.
[ERROR] src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java[94] (sizes) LineLength: Line is longer than 120 characters (found 135).
[ERROR] src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java[102] (sizes) LineLength: Line is longer than 120 characters (found 121).
[ERROR] src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java[104] (sizes) LineLength: Line is longer than 120 characters (found 132).
[ERROR] src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java[141] (sizes) LineLength: Line is longer than 120 characters (found 127).
[ERROR] src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceLinkListener.java[45] (sizes) LineLength: Line is longer than 120 characters (found 132).
[ERROR] src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceParameters.java[27] (sizes) LineLength: Line is longer than 120 characters (found 135).
[ERROR] src/main/java/org/apache/reef/io/network/NetworkConnectionService.java[36] (sizes) LineLength: Line is longer than 120 characters (found 138).
[ERROR] src/main/java/org/apache/reef/io/network/NetworkConnectionService.java[41] (sizes) LineLength: Line is longer than 120 characters (found 167).
[ERROR] src/main/java/org/apache/reef/io/network/NetworkConnectionService.java[42] (sizes) LineLength: Line is longer than 120 characters (found 149).

taegeonum commented:
@bgchun Uh oh... This is strange. I've merged with master and I don't have any errors in my local machine.

bgchun commented:
Can you fetch and merge again? Are you sure you have the latest checkstyle
update?

On Fri, Jul 10, 2015 at 11:42 PM, Tae-Geon Um <notifications@github.com>
wrote:

> @bgchun <https://github.com/bgchun> Uh oh... This is strange. I've merged
> with master and I don't have any errors in my local machine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255#issuecomment-120425800>
> .
>



-- 
Byung-Gon Chun

markusweimer commented:
You need to fetch from the apache git, not this mirror. The mirroring broke down again :-(

taegeonum commented:
@markusweimer  Got it. Thanks!

bgchun commented:
Commit changes that address only the errors. Don't commit the checkstyle
file.

On Fri, Jul 10, 2015 at 11:55 PM, Tae-Geon Um <notifications@github.com>
wrote:

> @markusweimer <https://github.com/markusweimer> Got it. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255#issuecomment-120428765>
> .
>



-- 
Byung-Gon Chun

bgchun commented:
The mirror's synchronized.

On Sat, Jul 11, 2015 at 12:00 AM, Byung-Gon Chun <bgchun@gmail.com> wrote:

> Commit changes that address only the errors. Don't commit the checkstyle
> file.
>
> On Fri, Jul 10, 2015 at 11:55 PM, Tae-Geon Um <notifications@github.com>
> wrote:
>
>> @markusweimer <https://github.com/markusweimer> Got it. Thanks!
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/incubator-reef/pull/255#issuecomment-120428765>
>> .
>>
>
>
>
> --
> Byung-Gon Chun
>



-- 
Byung-Gon Chun

taegeonum commented:
@bgchun I've addressed the checkstyle violation.

bgchun commented:
@taegeonum Fix the description in github and jira since we changed the interface and class name.

dafrista commented:
I think this is merged with commit 41ff1418781f885e4847e1c6386c40c9eeb30db8, but the PR didn't close automatically. @taegeonum can you close the PR manually?
